File: iosApp/iosApp/Modules/NFTDetail ✅/NFTDetailWireframe.swift
// Created by web3d4v on 27/05/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import UIKit
import web3walletcore

final class DefaultNFTDetailWireframe {
    private weak var parent: UIViewController?
    private let context: NFTDetailWireframeContext
    private let nftSendWireframeFactory: NFTSendWireframeFactory
    private let alertWireframeFactory: AlertWireframeFactory
    private let nftsService: NFTsService
    private let networksService: NetworksService

    private weak var vc: UIViewController?
    
    init(
        _ parent: UIViewController?,
        context: NFTDetailWireframeContext,
        nftSendWireframeFactory: NFTSendWireframeFactory,
        alertWireframeFactory: AlertWireframeFactory,
        nftsService: NFTsService,
        networksService: NetworksService
    ) {
        self.parent = parent
        self.context = context
        self.nftSendWireframeFactory = nftSendWireframeFactory
        self.alertWireframeFactory = alertWireframeFactory
        self.nftsService = nftsService
        self.networksService = networksService
    }
}

extension DefaultNFTDetailWireframe {
    
    func present() {
        let vc = wireUp()
        parent?.show(vc, sender: self)
    }
    
    func navigate(with destination: NFTDetailWireframeDestination) {
        if let input = destination as? NFTDetailWireframeDestination.Send {
            guard let network = networksService.network else { return }
            let context = NFTSendWireframeContext(network: network, nftItem: input.nft)
            nftSendWireframeFactory.make( vc, context: context).present()
        }
        if destination is NFTDetailWireframeDestination.Dismiss {
            vc?.popOrDismiss()
        }
    }
}

private extension DefaultNFTDetailWireframe {
    
    func wireUp() -> UIViewController {
        let vc: NFTDetailViewController = UIStoryboard(.nftDetail).instantiate()
        let interactor = DefaultNFTDetailInteractor(nftService: nftsService)
        let presenter = DefaultNFTDetailPresenter(
            view: WeakRef(referred: vc),
            wireframe: self,
            interactor: interactor,
            context: context
        )
        vc.presenter = presenter
        self.vc = vc
        return vc
    }
}

File: androidApp/src/main/java/com/sonsofcrypto/web3wallet/android/modules/compose/nftdetail/NFTDetailWireframe.kt
package com.sonsofcrypto.web3wallet.android.modules.compose.nftdetail

import androidx.fragment.app.Fragment
import com.sonsofcrypto.web3lib.services.networks.NetworksService
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3wallet.android.common.NavigationFragment
import com.sonsofcrypto.web3wallet.android.common.extensions.navigationFragment
import com.sonsofcrypto.web3wallet.android.common.ui.navigationFragment
import com.sonsofcrypto.web3wallet.android.modules.compose.confirmation.ConfirmationWireframeFactory
import com.sonsofcrypto.web3wallet.android.modules.compose.nftsend.NFTSendWireframeFactory
import com.sonsofcrypto.web3walletcore.modules.confirmation.ConfirmationWireframeContext
import com.sonsofcrypto.web3walletcore.modules.nftDetail.*
import com.sonsofcrypto.web3walletcore.modules.nftDetail.NFTDetailWireframeDestination.Back
import com.sonsofcrypto.web3walletcore.modules.nftDetail.NFTDetailWireframeDestination.Send
import com.sonsofcrypto.web3walletcore.modules.nftSend.NFTSendWireframeContext
import com.sonsofcrypto.web3walletcore.services.nfts.NFTsService

class DefaultNFTDetailWireframe(
    private val parent: WeakRef<Fragment>?,
    private val context: NFTDetailWireframeContext,
    private val nftsService: NFTsService,
    private val nftSendWireframeFactory: NFTSendWireframeFactory,
    private val networksService: NetworksService,
): NFTDetailWireframe {

    override fun present() {
        val fragment = wireUp()
        parent?.navigationFragment?.push(fragment, true)
    }

    override fun navigate(destination: NFTDetailWireframeDestination) {
        when (destination) {
            is Send -> {
                val network = networksService.network ?: return
                val context = NFTSendWireframeContext(
                    network = network,
                    nftItem = destination.nft
                )
                nftSendWireframeFactory.make(
                    parent = parent?.get(),
                    context = context
                ).present()
            }
            is Back -> { parent?.navigationFragment?.popOrDismiss() }
        }
    }

    private fun wireUp(): Fragment {
        val view = NFTDetailFragment()
        val interactor = DefaultNFTDetailInteractor(
            nftService = nftsService
        )
        val presenter = DefaultNFTDetailPresenter(
            WeakRef(view),
            this,
            interactor,
            context,
        )
        view.presenter = presenter
        return view
    }
}